### PR TITLE
Dataset resizing support

### PR DESCRIPTION
**Pull request by karjonas**
I took the liberty of cherry picking the code from @jrs65 branch. It should solve Issues #109 and #76.

### CODE REVIEW

**Comment by jrs65** [comment]
Oh sweet. I was planning on doing that later today as I finally have a little bit of unscheduled time! Thanks!

**Comment by karjonas** [comment]
What a coincidence I needed it myself today. Thanks for the code :)

**Comment by ferdonline** [comment]
Hi @karjonas & @jrs65  thanks for the contribution! 
Before I ping the admins, could you please add unit tests? Another thing, I guess we'd better not expose hdf5 types in the public api, which happens in createDataSet(..., hid_t). So to my mind either that method becomes protected or, eventually, they become a single method with an optional param.

**Comment by karjonas** [comment]
I guess what we need to do is to use either booleans, enums or a struct with settings in the C++ API that we map onto the corresponding HDF5 settings. I think this is a safe way to do it but we then need to figure out the possible settings we should allow. In this PR we can start with just the chunking setting and add the other settings @jrs65  wants in another PR.

**Comment by karjonas** [comment]
I have now rebased and updated this PR to use chunked dataset added by #113. I also added a unit test.

File: include/highfive/bits/H5DataSet_misc.hpp [hunk 1]
```diff
@@ -56,11 +56,31 @@ inline DataSpace DataSet::getMemSpace() const { return getSpace(); }
...
+
+    const size_t numDimensions = getSpace().getDimensions().size();
+    if (dims.size() != numDimensions) {
+        HDF5ErrMapper::ToException<DataSetException>(
+            "Invalid dataspace dimensions, got " + std::to_string(dims.size()) +
+            " expected " + std::to_string(numDimensions));
+    }
+
+    std::vector<hsize_t> real_dims(dims.size());
+    std::copy(dims.begin(), dims.end(), real_dims.begin());
```
Comment (hernando) [inline include/highfive/bits/H5DataSet_misc.hpp]: Use the constructor that takes a begin and end iterator.

File: include/highfive/bits/H5DataSet_misc.hpp [hunk 2]
```diff
@@ -56,11 +56,31 @@ inline DataSpace DataSet::getMemSpace() const { return getSpace(); }
...
+
+    if (H5Dset_extent(getId(), &(real_dims.at(0))) < 0) {
```
Comment (hernando) [inline include/highfive/bits/H5DataSet_misc.hpp]: .data() instead of &(realdims.at(0))

File: include/highfive/bits/H5Dataspace_misc.hpp [hunk 3]
```diff
@@ -35,6 +35,33 @@ inline DataSpace::DataSpace(const IT begin, const IT end) {
     }
 }
 
+inline DataSpace::DataSpace(const std::vector<size_t>& dims,
+                            const std::vector<size_t>& maxdims) {
+    std::vector<hsize_t> real_dims(dims.size());
+    std::vector<hsize_t> real_maxdims(maxdims.size());
+
+    if(dims.size() != maxdims.size()) {
```
Comment (hernando) [inline include/highfive/bits/H5Dataspace_misc.hpp]: spaces missing after keywords

File: include/highfive/bits/H5Dataspace_misc.hpp [hunk 4]
```diff
@@ -35,6 +35,33 @@ inline DataSpace::DataSpace(const IT begin, const IT end) {
...
+        throw DataSpaceException("dims and maxdims must be the same length.");
+    }
+
+    std::copy(dims.begin(), dims.end(), real_dims.begin());
```
Comment (hernando) [inline include/highfive/bits/H5Dataspace_misc.hpp]: Same comment as before about using the constructor

File: include/highfive/bits/H5Dataspace_misc.hpp [hunk 5]
```diff
@@ -92,6 +119,18 @@ inline std::vector<size_t> DataSpace::getDimensions() const {
     return details::to_vector_size_t(std::move(dims));
 }
 
+inline std::vector<size_t> DataSpace::getMaxDimensions() const {
+    std::vector<hsize_t> maxdims(getNumberDimensions());
+    if (H5Sget_simple_extent_dims(_hid, NULL, &(maxdims[0])) < 0) {
```
Comment (hernando) [inline include/highfive/bits/H5Dataspace_misc.hpp]: .data()

File: include/highfive/bits/H5Dataspace_misc.hpp [hunk 6]
```diff
@@ -92,6 +119,18 @@ inline std::vector<size_t> DataSpace::getDimensions() const {
...
+        HDF5ErrMapper::ToException<DataSetException>(
+            "Unable to get dataspace dimensions");
+    }
+
+    std::vector<size_t> res(maxdims.size());
+    std::copy(maxdims.begin(), maxdims.end(), res.begin());
```
Comment (hernando) [inline include/highfive/bits/H5Dataspace_misc.hpp]: use constructor

**Comment by karjonas** [comment]
The review issues are now fixed.

File: tests/unit/tests_high_five_base.cpp [hunk 7]
```diff
@@ -268,6 +264,63 @@ BOOST_AUTO_TEST_CASE(HighFiveGroupAndDataSet) {
...
+        // Resize the dataset to a larger size
+        dataset.resize({4, 6});
+
+        // Write into the new part of the dataset
+        dataset.select({3, 3}, {1, 3}).write(t2);
+
+        // Try resize out of bounds
+        BOOST_CHECK_THROW(dataset.resize({18, 1}), DataSetException);
+        // Try resize invalid dimensions
+        BOOST_CHECK_THROW(dataset.resize({1, 2, 3}), DataSetException);
```
Comment (hernando) [inline tests/unit/tests_high_five_base.cpp]: Add also this test to the creation of datasets with chunking if you don't mind.

File: tests/unit/tests_high_five_base.cpp [hunk 8]
```diff
@@ -242,8 +242,13 @@ BOOST_AUTO_TEST_CASE(HighFiveGroupAndDataSet) {
...
 
         // here we use the other signature
-        DataSet dataset_chuncked = file.createDataSet<float>(
+        DataSet dataset_chunked = file.createDataSet<float>(
             CHUNKED_DATASET_NAME, dataspace, goodChunking);
+
+        // Try resize out of bounds
+        BOOST_CHECK_THROW(dataset_chunked.resize({5, 7}), DataSetException);
+        // Try resize invalid dimensions
+        BOOST_CHECK_THROW(dataset_chunked.resize({1, 2, 3}), DataSetException);
```
Comment (hernando) [inline tests/unit/tests_high_five_base.cpp]: This is not what I meant. I meant to to try a bad chunking with more dimensions that the dataset. You've already provided a resizing test elsewhere.

**Comment by hernando** [comment]
Can you rename the issue to "Dataset resizing support"?

File: tests/unit/tests_high_five_base.cpp (hunk 8 above)
Comment (karjonas) [reply tests/unit/tests_high_five_base.cpp]: I see. Maybe also test a dataset resize that is smaller than the chunking size?

Comment (hernando) [reply tests/unit/tests_high_five_base.cpp]: That wouldn't hurt for sure.

**Comment by hernando** [comment]
You have to add @tristan0x as reviewer as I can't approve the PR.

File: src/examples/create_extensible_dataset.cpp [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        // Write into the initial part of the dataset
+        double t1[3][1] = {{2.0}, {2.0}, {4.0}};
+        dataset.select({0, 0}, {3, 1}).write(t1);
+
+        // Resize the dataset to a larger size
+        dataset.resize({4, 6});
+
+        // Write into the new part of the dataset
+        double t2[1][3] = {{4.0, 8.0, 6.0}};
+        dataset.select({3, 3}, {1, 3}).write(t2);
```
Comment (tristan0x) [inline src/examples/create_extensible_dataset.cpp]: Can you please check that data has been written as expected?

File: src/examples/create_extensible_dataset.cpp [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+// Create a dataset name "dset" of double 4x6
+//
+int main(void) {
+    using namespace HighFive;
+    try {
+        // Create a new file using the default property lists.
+        File file(FILE_NAME, File::ReadWrite | File::Create | File::Truncate);
+
+        // Create a dataspace with intial shape and max shape
```
Comment (tristan0x) [inline src/examples/create_extensible_dataset.cpp]: in**i**tial

File: src/examples/create_extensible_dataset.cpp [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        dataset.select({0, 0}, {3, 1}).write(t1);
+
+        // Resize the dataset to a larger size
+        dataset.resize({4, 6});
+
+        // Write into the new part of the dataset
+        double t2[1][3] = {{4.0, 8.0, 6.0}};
+        dataset.select({3, 3}, {1, 3}).write(t2);
+
+    } catch (Exception& err) {
```
Comment (tristan0x) [inline src/examples/create_extensible_dataset.cpp]: `err` is const.

File: tests/unit/tests_high_five_base.cpp [hunk 12]
```diff
@@ -268,6 +269,63 @@ BOOST_AUTO_TEST_CASE(HighFiveGroupAndDataSet) {
...
+
+        // Create the dataset
+        DataSet dataset = file.createDataSet(DATASET_NAME, dataspace,
+                                             AtomicType<double>(), props);
+
+        // Write into the initial part of the dataset
+        dataset.select({0, 0}, {3, 1}).write(t1);
+
+        // Resize the dataset to a larger size
+        dataset.resize({4, 6});
```
Comment (tristan0x) [inline tests/unit/tests_high_five_base.cpp]: Can you also check that new reported size is correct?

**Comment by karjonas** [comment]
Review issues yet again fixed.

**Comment by tristan0x** [comment]
I bother you one more time @karjonas : can you please squash your changes before I merge?

**Comment by karjonas** [comment]
That should do it.

**Comment by tristan0x** [comment]
Thank you @karjonas

**Pull request merged**